client/mysqlbinlog.cc

```diff
@@ -2948,7 +2948,6 @@ static Exit_status dump_local_log_entries(PRINT_EVENT_INFO *print_event_info,
       if (((longlong)stop_position != stop_position_default) &&
           stop_position > my_b_tell(file))
       {
-          retval = OK_STOP;
```

Comment:
In looking to answer your questions 1 & 2, I was going through the old PR's discussion around this ([here](https://github.com/MariaDB/server/pull/3400#discussion_r1678494313), for reference); and I'm thinking we should go with option 2 actually:
> Move these error messages to happen after the dump_log_entries() loop, and cache the last read stop_position and stop_datetime (when applicable) to compare against

because it
  1) is more general (i.e. it will apply to `dump_remote_log_entries()` as well, though we should extend the test with that. Though I can take it on, and the current test I provided is fine enough for you to get started on coding (although you're welcome to take it on as well, if you'd prefer)), 
  2) is more consistent with where one would expect to find the warning (i.e. a warning about last thing processed would make sense to be checked only after all logs have been processed, rather than after _each_ log has been processed, and
  3) lessens the number of conditionals (always good :smile:)

Note the reason this wasn't done initially, is that the original thought was to use condition `retval != OK_STOP`; however, `retval` can be changed after it is set to `OK_STOP` (at the end of `dump_local_log_entries()`, so the warning might be swallowed. But if we cache the last binlog position and event time, then we won't worry about the `retval` status being swallowed).

Then to your third question: correct, let's just have them in the same file for consistency.

Reply:
hi @bnestere , that makes sense. 

I will go ahead to make changes in ```dump_local_log_entries``` to keep tracking last position and datetime with 2 new global variables. And after the ```dump_log_entries``` loop, decide if need to emit the warning based on those and requested ```--stop_position``` and ```--stop_datetime```. As you said, your test cases will help validate the change.

For ```dump_remote_log_entries```, it is not that clear for me at which point we should track last position and datetime after reading code, and not sure how to test, so I will not modify that for now, but I will make sure for the check after ```dump_log_entries``` loop, it will not emit warning for ```dump_remote_log_entries```, which is same as current implementation.

Reply:
Thanks @XiaochuanCui ! 

As to 
> For dump_remote_log_entries, it is not that clear for me at which point we should track last position and datetime after reading code

I think both local and remote can use the same spot. That is, in `process_event()`, at the `end` label:
```
end:
  rec_count++;
  // here
```

Then here are updated test files that should verify both local and remote behavior. All I did was split the actual validation checks into a `.inc` file with a parameter `use_local_logs`, which sets mysqlbinlog up to either use local logs, or use `--read-from-remote-server`.

`mysql-test/suite/rpl/t/binlog_mysqlbinlog_warn_stop_position.test`:
```
#
#   Test ensures that stop conditions (--stop-position and --stop-datetime)
# work correctly for mysqlbinlog. This high level test sets up the binary log
# (and tags certain locations for comparison), and the helper file
# binlog_mysqlbinlog_warn_stop_position.inc performs the actual tests.
#
# References:
#   MDEV-27037: mysqlbinlog emits a warning when reaching EOF before
#               stop-condition
#   MDEV-35528: mariadb-binlog cannot process more than 1 logfiles when
#               --stop-datetime is specified
#
--source include/have_binlog_format_row.inc

--let $binlog_out_relpath=tmp/warn_pos_test_file.out
--let $binlog_out= $MYSQLTEST_VARDIR/$binlog_out_relpath

--let $b1_timestamp1= 2024-12-01 10:20:30.123456
--let $b1_timestamp2= 2024-12-02 10:20:30.123456
--let $b1_timestamp3= 2024-12-03 10:20:30.123456
--let $b2_timestamp1= 2024-12-04 10:20:30.123456
--let $b2_timestamp2= 2024-12-05 10:20:30.123456
--let $b2_timestamp3= 2024-12-06 10:20:30.123456
--let $b2_timestamp_not_reached= 2024-12-07 10:20:30.123456
--eval SET TIMESTAMP=UNIX_TIMESTAMP('$b1_timestamp1')
--let $binlog_f1=  query_get_value(SHOW MASTER STATUS, File, 1)
create table t1 (a int);
insert into t1 values (1);
--eval SET TIMESTAMP=UNIX_TIMESTAMP('$b1_timestamp2')
insert into t1 values (2);
--eval SET TIMESTAMP=UNIX_TIMESTAMP('$b1_timestamp3')
--let $binlog_f1_pre_rotate= query_get_value(SHOW MASTER STATUS, Position, 1)
flush binary logs;
--let $binlog_f2=  query_get_value(SHOW MASTER STATUS, File, 1)
--eval SET TIMESTAMP=UNIX_TIMESTAMP('$b2_timestamp1')
insert into t1 values (3);
--echo # Tag binlog_f2_mid
--let $binlog_f2_mid= query_get_value(SHOW MASTER STATUS, Position, 1)
insert into t1 values (4);
--eval SET TIMESTAMP=UNIX_TIMESTAMP('$b2_timestamp2')
--let $binlog_f2_gtid_after_midpoint= `SELECT @@GLOBAL.gtid_binlog_pos`
insert into t1 values (5);
insert into t1 values (6);
insert into t1 values (7);
--let $binlog_f2_last_gtid= `SELECT @@GLOBAL.gtid_binlog_pos`
--let $binlog_f2_pre_rot= query_get_value(SHOW MASTER STATUS, Position, 1)
--eval SET TIMESTAMP=UNIX_TIMESTAMP('$b2_timestamp3')
flush binary logs;
drop table t1;


--echo # Ensuring binary log order is correct
--let $binlog_f1_show= query_get_value(SHOW BINARY LOGS, Log_name, 1)
--let $binlog_f1_end= query_get_value(SHOW BINARY LOGS, File_size, 1)
if (`SELECT strcmp('$binlog_f1','$binlog_f1_show') != 0`)
{
  --echo # Real binlog_f1: $binlog_f1
  --echo # First binlog in SHOW BINLOG FILES: $binlog_f1_show
  --die Wrong order of binary log files in SHOW BINARY LOGS
}
--let $binlog_f2_show= query_get_value(SHOW BINARY LOGS, Log_name, 2)
--let $binlog_f2_end= query_get_value(SHOW BINARY LOGS, File_size, 2)
if (`SELECT strcmp('$binlog_f2','$binlog_f2_show') != 0`)
{
  --echo # Real binlog_f2: $binlog_f2
  --echo # First binlog in SHOW BINLOG FILES: $binlog_f2_show
  --die Wrong order of binary log files in SHOW BINARY LOGS
}

--echo # Ensuring file offset of binlog_f2_mid < binlog_f1_end (see test 1.d)
if ($binlog_f2_mid > $binlog_f1_end)
{
  --echo # Binlog 1 end: $binlog_f1:$binlog_f1_end
  --echo # Binlog 2 stop point: $binlog_f2:$binlog_f2_mid
  --die Mid point chosen to end in binlog 2 does not exist in earlier binlog
}

--echo #
--echo #
--echo # Test using --read-from-remote-server
--echo #
--let $use_local_files= 0
--source binlog_mysqlbinlog_warn_stop_position.inc

--echo #
--echo #
--echo # Test using local binlog files
--echo #
--let $use_local_files= 1
--source binlog_mysqlbinlog_warn_stop_position.inc

--echo # End of binlog_mysqlbinlog_warn_stop_position.test

```

And its helper file 
`mysql-test/suite/rpl/t/binlog_mysqlbinlog_warn_stop_position.inc`:
```
#
# Helper file that ensures mysqlbinlog --stop-position and --stop-datetime
# behavior for local files or a remote server
#
# Parameters:
#   use_local_files (bool): A boolean that changes which source to use for
#     mysqlbinlog. When true, uses local files; when false, reads remotely.
#

--connection default
--let $MYSQLD_DATADIR= `select @@datadir`

# PARAM_READ_FROM_REMOTE is used as a parameter to mysqlbinlog (_OUT suffix is
# output in echo commands). If using local files, they are blank; if reading
# from remote server, it is overridden to the correct values.
--let $PARAM_READ_FROM_REMOTE=
# Used in echo statements to remove potentially changing values
--let $PARAM_READ_FROM_REMOTE_OUT=
if (!$use_local_files)
{
  --let $PARAM_READ_FROM_REMOTE= --read-from-remote-server --user=root --host=127.0.0.1 --port=$MASTER_MYPORT
  --let $PARAM_READ_FROM_REMOTE_OUT= --read-from-remote-server
  # binlog files in --read-from-remote-server don't use file system path
  --let $binlog_f1_full= $binlog_f1
  --let $binlog_f2_full= $binlog_f2
}
if ($use_local_files)
{
  # If using local files, file system path to the binlog files is needed
  --let $binlog_f1_full= $MYSQLD_DATADIR/$binlog_f1
  --let $binlog_f2_full= $MYSQLD_DATADIR/$binlog_f2
}

--echo #
--echo # --stop-position tests
--echo #

--echo #  Case 1.a) With one binlog file, a --stop-position before the end of
--echo # the file should not result in a warning
--echo # MYSQL_BINLOG $PARAM_READ_FROM_REMOTE_OUT --stop-position=binlog_f1_pre_rotate binlog_f1_full --result-file=$binlog_out_relpath 2>&1
--exec $MYSQL_BINLOG $PARAM_READ_FROM_REMOTE --stop-position=$binlog_f1_pre_rotate $binlog_f1_full --result-file=$binlog_out 2>&1

--echo #  Case 1.b) With one binlog file, a --stop-position at the exact end of
--echo # the file should not result in a warning
--echo # MYSQL_BINLOG $PARAM_READ_FROM_REMOTE_OUT --stop-position=binlog_f1_end binlog_f1_full --result-file=$binlog_out_relpath 2>&1
--exec $MYSQL_BINLOG $PARAM_READ_FROM_REMOTE --stop-position=$binlog_f1_end $binlog_f1_full --result-file=$binlog_out 2>&1

--echo #  Case 1.c) With one binlog file, a --stop-position past the end of the
--echo # file should(!) result in a warning
--let $binlog_f1_over_eof= `SELECT $binlog_f1_end + 10`
--echo # MYSQL_BINLOG $PARAM_READ_FROM_REMOTE_OUT --short-form --stop-position=binlog_f1_over_eof binlog_f1_full --result-file=$binlog_out_relpath 2>&1
--replace_result $binlog_f1_over_eof <BINLOG_F1_OVER_EOF>
--exec $MYSQL_BINLOG $PARAM_READ_FROM_REMOTE --short-form --stop-position=$binlog_f1_over_eof $binlog_f1_full --result-file=$binlog_out 2>&1

--echo #  Case 1.d) With two binlog files, a --stop-position targeting b2 which
--echo # exists in the size of b1 should:
--echo #    1) not provide any warnings
--echo #    2) not prevent b2 from outputting its desired events before the
--echo #       stop position
--echo # MYSQL_BINLOG $PARAM_READ_FROM_REMOTE_OUT --stop-position=binlog_f2_mid binlog_f1_full binlog_f2_full --result-file=$binlog_out_relpath 2>&1
--exec $MYSQL_BINLOG $PARAM_READ_FROM_REMOTE --stop-position=$binlog_f2_mid $binlog_f1_full $binlog_f2_full --result-file=$binlog_out 2>&1

--let $server_id= `SELECT @@GLOBAL.server_id`
--let $domain_id= `SELECT @@GLOBAL.gtid_domain_id`
--let $assert_file= $binlog_out
--let $assert_text= Ensure all intended GTIDs are present
--let $assert_select= GTID $domain_id-$server_id-
--let $assert_count= 4
--source include/assert_grep.inc

--let $assert_text= Ensure the next GTID binlogged is _not_ present
--let $assert_select= GTID $binlog_f2_gtid_after_midpoint
--let $assert_count= 0
--source include/assert_grep.inc

--echo #  Case 1.e) With two binlog files, a --stop-position targeting the end
--echo # of binlog 2 should:
--echo #    1) not provide any warnings
--echo #    2) not prevent b2 from outputting its entire binary log
--echo # MYSQL_BINLOG $PARAM_READ_FROM_REMOTE_OUT --stop-position=binlog_f2_end binlog_f1_full binlog_f2_full --result-file=$binlog_out_relpath 2>&1
--exec $MYSQL_BINLOG $PARAM_READ_FROM_REMOTE --stop-position=$binlog_f2_end $binlog_f1_full $binlog_f2_full --result-file=$binlog_out 2>&1

--let $server_id= `SELECT @@GLOBAL.server_id`
--let $domain_id= `SELECT @@GLOBAL.gtid_domain_id`
--let $assert_text= Ensure a GTID exists for each transaction
--let $assert_select= GTID $domain_id-$server_id-
--let $assert_count= 8
--source include/assert_grep.inc

--let $assert_text= Ensure the last GTID binlogged is present
--let $assert_select= GTID $binlog_f2_last_gtid
--let $assert_count= 1
--source include/assert_grep.inc

--echo #  Case 1.f) With two binlog files, a --stop-position targeting beyond
--echo # the eof of binlog 2 should:
--echo #    1) provide a warning that the stop position was not reached
--echo #    2) not prevent b2 from outputting its entire binary log
--let $binlog_f2_over_eof= `SELECT $binlog_f2_end + 10`
--echo # MYSQL_BINLOG $PARAM_READ_FROM_REMOTE_OUT --stop-position=binlog_f2_over_eof binlog_f1_full binlog_f2_full --result-file=$binlog_out_relpath 2>&1
--exec $MYSQL_BINLOG $PARAM_READ_FROM_REMOTE --stop-position=$binlog_f2_over_eof $binlog_f1_full $binlog_f2_full --result-file=$binlog_out 2>&1

--let $server_id= `SELECT @@GLOBAL.server_id`
--let $domain_id= `SELECT @@GLOBAL.gtid_domain_id`
--let $assert_text= Ensure a GTID exists for each transaction
--let $assert_select= GTID $domain_id-$server_id-
--let $assert_count= 8
--source include/assert_grep.inc


--echo #
--echo # --stop-datetime tests
--echo # Note: MDEV-35528 reported that mysqlbinlog would fail on tests cases
--echo # 2.d, 2.e, and 2.f.
--echo #

--echo #  Case 2.a) With one binlog file, a --stop-datetime before the end of
--echo # the file should not result in a warning
--echo # MYSQL_BINLOG $PARAM_READ_FROM_REMOTE_OUT --stop-datetime='$b1_timestamp2' binlog_f1_full --result-file=$binlog_out_relpath 2>&1
--exec $MYSQL_BINLOG $PARAM_READ_FROM_REMOTE --stop-datetime='$b1_timestamp2' $binlog_f1_full --result-file=$binlog_out 2>&1

--echo #  Case 2.b) With one binlog file, a --stop-datetime at the end of the
--echo # file should not result in a warning
--echo # MYSQL_BINLOG $PARAM_READ_FROM_REMOTE_OUT --stop-datetime='$b1_timestamp3' binlog_f1_full --result-file=$binlog_out_relpath 2>&1
--exec $MYSQL_BINLOG $PARAM_READ_FROM_REMOTE --stop-datetime='$b1_timestamp3' $binlog_f1_full --result-file=$binlog_out 2>&1

--echo #  Case 2.c) With one binlog file, a --stop-datetime beyond the end of
--echo # the file should(!) result in a warning
--echo # MYSQL_BINLOG $PARAM_READ_FROM_REMOTE_OUT --stop-datetime='$b2_timestamp1' binlog_f1_full --result-file=$binlog_out_relpath 2>&1
--exec $MYSQL_BINLOG $PARAM_READ_FROM_REMOTE --stop-datetime='$b2_timestamp1' $binlog_f1_full --result-file=$binlog_out 2>&1

--echo #  Case 2.d) With two binlog files, a --stop-datetime within the
--echo # timespan of binlog 2 should:
--echo #    1) not provide any warnings
--echo #    2) not prevent binlog 1 or 2 from outputting the desired events

--echo # MYSQL_BINLOG $PARAM_READ_FROM_REMOTE_OUT --stop-datetime='$b2_timestamp2' binlog_f1_full binlog_f2_full --result-file=$binlog_out_relpath 2>&1
--exec $MYSQL_BINLOG $PARAM_READ_FROM_REMOTE --stop-datetime='$b2_timestamp2' $binlog_f1_full $binlog_f2_full --result-file=$binlog_out 2>&1

--let $assert_text= Ensure all intended GTIDs are present
--let $assert_select= GTID $domain_id-$server_id-
--let $assert_count= 4
--source include/assert_grep.inc

--let $assert_text= Ensure the next GTID binlogged is _not_ present
--let $assert_select= GTID $binlog_f2_gtid_after_midpoint
--let $assert_count= 0
--source include/assert_grep.inc

--echo #  Case 2.e) With two binlog files, a --stop-datetime at the end of
--echo # binlog 2 should:
--echo #    1) not provide any warnings
--echo #    2) not prevent binlog 1 or 2 from outputting all events
--echo # MYSQL_BINLOG $PARAM_READ_FROM_REMOTE_OUT --stop-datetime='$b2_timestamp3' binlog_f1_full binlog_f2_full --result-file=$binlog_out_relpath 2>&1
--exec $MYSQL_BINLOG $PARAM_READ_FROM_REMOTE --stop-datetime='$b2_timestamp3' $binlog_f1_full $binlog_f2_full --result-file=$binlog_out 2>&1

--let $assert_text= Ensure a GTID exists for each transaction
--let $assert_select= GTID $domain_id-$server_id-
--let $assert_count= 8
--source include/assert_grep.inc

--echo #  Case 2.f) With two binlog files, a --stop-datetime beyond the end of
--echo # binlog 2 should:
--echo #    1) provide a warning that the stop datetime was not reached
--echo #    2) not prevent binlog 1 or 2 from outputting all events
--echo # MYSQL_BINLOG $PARAM_READ_FROM_REMOTE_OUT --stop-datetime='$b2_timestamp3' binlog_f1_full binlog_f2_full --result-file=$binlog_out_relpath 2>&1
--exec $MYSQL_BINLOG $PARAM_READ_FROM_REMOTE --stop-datetime='$b2_timestamp3' $binlog_f1_full $binlog_f2_full --result-file=$binlog_out 2>&1

--let $assert_text= Ensure a GTID exists for each transaction
--let $assert_select= GTID $domain_id-$server_id-
--let $assert_count= 8
--source include/assert_grep.inc

```

Reply:
> ```
> --source include/have_binlog_format_row.inc
> ```

Why don’t we test all binlog formats?

Reply:
Hi @bnestere, good point for tracking last position and datetime at the end of ```process_event()```. 

I thought we cannot do this since I noticed there is a branch ```handle_event_raw_mode``` in ```dump_remote_log_entries``` which does not call ```process_event()```. If that is not a problem. I will move the tracking logic to ```process_event()``` as your suggested.

Thank you for writing the test cases. I will test with them and also check if we can test other format as @ParadoxV5 suggested.

Reply:
@XiaochuanCui Right, it is not a problem in raw mode. `--stop-position` and `--stop-datetime` are not actually used in raw mode (which our [documentation](https://mariadb.com/kb/en/mariadb-binlog-options/) also is consistent with).

Then for @ParadoxV5 's question
> Why don’t we test all binlog formats?

`--stop-position` and `--stop-datetime` _should_ be binlog format independent, but it doesn't hurt to verify :) And I don't think the test result is reliant on any specific binlog offset values, so it shouldn't need any additional modification to work for all, other than changing 
```diff
---source include/have_binlog_format_row.inc
+--source include/have_log_bin.inc
```

Reply:
Hi @bnestere , it seems the code to control timestamp in binlogs such as ```--eval SET TIMESTAMP=UNIX_TIMESTAMP('$b1_timestamp1')``` does not work as expected. 

This is same for the existing [binlog_mysqlbinlog_warn_stop_datetime.test](https://github.com/MariaDB/server/blob/mariadb-10.5.27/mysql-test/suite/binlog/t/binlog_mysqlbinlog_warn_stop_datetime.test). In last test case, it set "Stop datetime after EOF" as '2035-01-19 03:14:05'. It meant to be anytime after '2001-09-08 21:46:55', but when it is set to sometime before the test run such as '2023-09-08 21:46:55', it will not emit warning. If set to sometime after the test run, it will work. This implies the timestamps in binlog are still set to current time.

Checked the binlog created by test, before the events created by the test case, there are some other events with timestamp as the time when test run.
```
# ./build/client/mysqlbinlog ./build/mysql-test/var/mysqld.1/data/master-bin.000001
/*!50530 SET @@SESSION.PSEUDO_SLAVE_MODE=1*/;
/*!40019 SET @@session.max_insert_delayed_threads=0*/;
/*!50003 SET @OLD_COMPLETION_TYPE=@@COMPLETION_TYPE,COMPLETION_TYPE=0*/;
DELIMITER /*!*/;
# at 4
#241213 18:56:27 server id 1  end_log_pos 256 CRC32 0xec3ed765 	Start: binlog v 4, server v 10.5.28-MariaDB-log created 241213 18:56:27 at startup
ROLLBACK/*!*/;
BINLOG '
W4NcZw8BAAAA/AAAAAABAAAAAAQAMTAuNS4yOC1NYXJpYURCLWxvZwAAAAAAAAAAAAAAAAAAAAAA
AAAAAAAAAAAAAAAAAABbg1xnEzgNAAgAEgAEBAQEEgAA5AAEGggAAAAICAgCAAAACgoKAAAAAAAA
AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA
AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA
AAAAAAAAAAAEEwQADQgICAoKCgFl1z7s
'/*!*/;
# at 256
#241213 18:56:27 server id 1  end_log_pos 285 CRC32 0x22b777f8 	Gtid list []
# at 285
#241213 18:56:27 server id 1  end_log_pos 329 CRC32 0xaae30c17 	Binlog checkpoint master-bin.000001
# at 329
#010909  1:46:40 server id 1  end_log_pos 371 CRC32 0xcfe9d22e 	GTID 0-1-1 ddl
/*M!100101 SET @@session.skip_parallel_replication=0*//*!*/;
/*M!100001 SET @@session.gtid_domain_id=0*//*!*/;
/*M!100001 SET @@session.server_id=1*//*!*/;
/*M!100001 SET @@session.gtid_seq_no=1*//*!*/;
# at 371
#010909  1:46:40 server id 1  end_log_pos 471 CRC32 0xbec6e329 	Query	thread_id=5	exec_time=734116188	error_code=0
use `test`/*!*/;
SET TIMESTAMP=1000000000/*!*/;
SET @@session.pseudo_thread_id=5/*!*/;
SET @@session.foreign_key_checks=1, @@session.sql_auto_is_null=0, @@session.unique_checks=1, @@session.autocommit=1, @@session.check_constraint_checks=1, @@session.sql_if_exists=0, @@session.explicit_defaults_for_timestamp=0/*!*/;
SET @@session.sql_mode=1411383296/*!*/;
SET @@session.auto_increment_increment=1, @@session.auto_increment_offset=1/*!*/;
/*!\C latin1 *//*!*/;
SET @@session.character_set_client=latin1,@@session.collation_connection=8,@@session.collation_server=8/*!*/;
SET @@session.lc_time_names=0/*!*/;
SET @@session.collation_database=DEFAULT/*!*/;
CREATE TABLE t1(word VARCHAR(20))
/*!*/;
# at 471
#010909  1:46:50 server id 1  end_log_pos 513 CRC32 0xe9181c82 	GTID 0-1-2
/*M!100001 SET @@session.gtid_seq_no=2*//*!*/;
START TRANSACTION
/*!*/;
# at 513
#010909  1:46:50 server id 1  end_log_pos 614 CRC32 0x9d38717e 	Query	thread_id=5	exec_time=734116178	error_code=0
SET TIMESTAMP=1000000010/*!*/;
INSERT INTO t1 VALUES ("abirvalg")
/*!*/;
# at 614
#010909  1:46:50 server id 1  end_log_pos 687 CRC32 0x55eb6bca 	Query	thread_id=5	exec_time=734116178	error_code=0
SET TIMESTAMP=1000000010/*!*/;
COMMIT
/*!*/;
# at 687
#010909  1:47:00 server id 1  end_log_pos 729 CRC32 0xc5e28b7c 	GTID 0-1-3
/*M!100001 SET @@session.gtid_seq_no=3*//*!*/;
START TRANSACTION
/*!*/;
# at 729
#010909  1:47:00 server id 1  end_log_pos 827 CRC32 0x9b75ead8 	Query	thread_id=5	exec_time=734116168	error_code=0
SET TIMESTAMP=1000000020/*!*/;
INSERT INTO t1 SELECT * FROM t1
/*!*/;
# at 827
#010909  1:47:00 server id 1  end_log_pos 900 CRC32 0x2ff04555 	Query	thread_id=5	exec_time=734116168	error_code=0
SET TIMESTAMP=1000000020/*!*/;
COMMIT
/*!*/;
# at 900
#010909  1:47:00 server id 1  end_log_pos 948 CRC32 0x3122af59 	Rotate to master-bin.000002  pos: 4
DELIMITER ;
# End of log file
ROLLBACK /* added by mysqlbinlog */;
/*!50003 SET COMPLETION_TYPE=@OLD_COMPLETION_TYPE*/;
/*!50530 SET @@SESSION.PSEUDO_SLAVE_MODE=0*/;
```
So, when running mysqlbinlog with ```--stop-datetime``` set to sometime before now, it will not return any events, because the first several events are later than ```--stop-datetime```, so it skips all later events even they are before ```--stop-datetime```.

```
# ./build/client/mysqlbinlog ./build/mysql-test/var/mysqld.1/data/master-bin.000001 --stop-datetime='2015-01-19 03:14:05'
/*!50530 SET @@SESSION.PSEUDO_SLAVE_MODE=1*/;
/*!40019 SET @@session.max_insert_delayed_threads=0*/;
/*!50003 SET @OLD_COMPLETION_TYPE=@@COMPLETION_TYPE,COMPLETION_TYPE=0*/;
DELIMITER /*!*/;
DELIMITER ;
# End of log file
ROLLBACK /* added by mysqlbinlog */;
/*!50003 SET COMPLETION_TYPE=@OLD_COMPLETION_TYPE*/;
/*!50530 SET @@SESSION.PSEUDO_SLAVE_MODE=0*/;
```

I will try to find a way to fix the test case. Please let me know if you have any idea. Thanks.

Reply:
> So, when running mysqlbinlog with `--stop-datetime` set to sometime before now, it will not return any events, because the first several events are later than `--stop-datetime`, so it skips all later events even they are before `--stop-datetime`.

👀
This means `--stop-datetime` needs to be reïmplemented as a filter similar to the different domains of `--stop-position` GTIDs.

This would turn a bugfix into a new feature, though.
* If we don’t mind staying on 10.5, we can simply include this phenomenon to MDEV-35528’s title and description instead of recording a separate bug entry.
  * It ain’t broke, but do we still fix `--stop_position` then?
* If we’d it a separate topic, let’s just figure out a workaround to patch these tests with.

Reply:
> This means ```--stop-datetime``` needs to be reïmplemented as a filter similar to the different domains of ```--stop-position``` GTIDs. This would turn a bugfix into a new feature, though.

Exactly, if anyone ```SET TIMESTAMP``` to some time in future, and call mariadb-binlog with ```--stop-datetime``` specified as current time, then it will stop at the event with future time and skip subsequent events.  The implementation assumes all events in binlog are in chronological order. 
I can create a separate issue for this with more information given the scope is much bigger, and we may need to evaluate if this is a common situation and if need to fix.

> It ain’t broke, but do we still fix ```--stop-position``` then?       

It is true that we do not need to change anything for ```--stop-position```, but maybe better to make the logic for ```--stop-datetime``` and ```--stop-position``` consistent.

> If we’d it a separate topic, let’s just figure out a workaround to patch these tests with.

To fix the test case, one workaround is to set all preset timestamp as some time in future. That way, the test will not be impacted by additional events at beginning of binlogs.

Reply:
> This would turn a bugfix into a new feature, though.

Indeed.  Though..

> If we don’t mind staying on 10.5, we can simply include this phenomenon to MDEV-35528’s title and description instead of recording a separate bug entry.

I think we should file the new functionality as a new JIRA ticket to fix separately (and in a non-GA release). I imagine we will want some discussion about it (and it would be easier to keep focused). 

For now, let's just fix the test by making the binary log header events consistent with the rest of the events using `RESET MASTER` (which clears the existing binary log state, and starts fresh using the `timestamp` variable):

```diff

 --let $b1_timestamp1= 2024-12-01 10:20:30.123456
 --let $b1_timestamp2= 2024-12-02 10:20:30.123456
 --let $b1_timestamp3= 2024-12-03 10:20:30.123456
 --let $b2_timestamp1= 2024-12-04 10:20:30.123456
 --let $b2_timestamp2= 2024-12-05 10:20:30.123456
 --let $b2_timestamp3= 2024-12-06 10:20:30.123456
 --let $b2_timestamp_not_reached= 2024-12-07 10:20:30.123456
 --eval SET TIMESTAMP=UNIX_TIMESTAMP('$b1_timestamp1')
+RESET MASTER;
 --let $binlog_f1=  query_get_value(SHOW MASTER STATUS, File, 1)
 create table t1 (a int);
```

Reply:
Hi @bnestere , a new revision of changes is updated for this PR as we discussed. Please take a look when you get a chance.

One difference with our discussion is that the change only handles ```--stop-datetime```, while leave all existing logic for ```--stop-position``` as it is (Tried to also record last processed event position at same place in process_event method, but for some reason, it can only record the second last position in each file). Thus for ```--stop-position```, emitting warning for log files on remote server is not supported now. 
Another difference is that it separated MTR for ```--stop-datetime``` and ```--stop-position``` to make it clearer.

Also, a new [JIRA issue](https://jira.mariadb.org/browse/MDEV-35664
)  is created for the other mariadb-binlog improvement out of scope of this bug fix.

Reply:
🤔 the new filter design would supercede the current `--stop-datetime` approach though.
Nonetheless, this contribution is appreciated 💪, especially that bnestere’s thinking about
> […] to fix separately (and in a non-GA release).

Reply:
Thanks @XiaochuanCui ! 

To the inconsistency with `--stop-position`, the issue that (I imagine) you saw was that you were storing `pos` as the last processed position, which points to the starting offset of the event in the binary log; however, the `--stop-position` comparison needs to compare against the ending offset. That allows for the `--stop-position` to specify any offset that exists within that event, and the logic would still work (without providing a warning). So the last processed position needs to be the end of the event. 

As Otto mentions, your PR indeed fixes the regression, and I am happy to accept it as-is. Though it will temporarily leave the code in an inconsistent state for where the warnings are issued. I've created a git patch which you are welcome to include in your commit to make everything consistent at once (with a few additional formatting fixes, that I'll otherwise mention in one-off review comments). Otherwise, I"ll just apply it later.

Can you please also update your git commit message mentioning myself as the reviewer? 
```
Reviewed by:
Brandon Nesterenko <brandon.nesterenko@mariadb.com>
```

```diff
diff --git a/client/mysqlbinlog.cc b/client/mysqlbinlog.cc
index 2874b3e881b..c6c6325873e 100644
--- a/client/mysqlbinlog.cc
+++ b/client/mysqlbinlog.cc
@@ -143,6 +143,7 @@ static uint verbose= 0;
 
 static ulonglong start_position, stop_position;
 static const longlong stop_position_default= (longlong)(~(my_off_t)0);
+static ulonglong last_processed_position= 0;
 #define start_position_mot ((my_off_t)start_position)
 #define stop_position_mot  ((my_off_t)stop_position)
 
@@ -997,6 +998,8 @@ static bool print_row_event(PRINT_EVENT_INFO *print_event_info, Log_event *ev,
   determining how to print.
   @param[in] ev Log_event to process.
   @param[in] pos Offset from beginning of binlog file.
+  @param[in] ev_end_pos Offset that represents the end of the event in the
+                        binlog file.
   @param[in] logname Name of input binlog.
 
   @retval ERROR_STOP An error occurred - the program should terminate.
@@ -1005,7 +1008,7 @@ static bool print_row_event(PRINT_EVENT_INFO *print_event_info, Log_event *ev,
   events to process has been reached and the program should terminate.
 */
 Exit_status process_event(PRINT_EVENT_INFO *print_event_info, Log_event *ev,
-                          my_off_t pos, const char *logname)
+                          my_off_t pos, my_off_t ev_end_pos, const char *logname)
 {
   char ll_buff[21];
   Log_event_type ev_type= ev->get_type_code();
@@ -1461,7 +1464,8 @@ Exit_status process_event(PRINT_EVENT_INFO *print_event_info, Log_event *ev,
   retval= ERROR_STOP;
 end:
   rec_count++;
-  last_processed_datetime = last_ev_when;
+  last_processed_datetime= last_ev_when;
+  last_processed_position= ev_end_pos;
 
   DBUG_PRINT("info", ("end event processing"));
   /*
@@ -2379,7 +2383,8 @@ static Exit_status handle_event_text_mode(PRINT_EVENT_INFO *print_event_info,
       if (old_off != BIN_LOG_HEADER_SIZE)
         *len= 1;         // fake event, don't increment old_off
     }
-    Exit_status retval= process_event(print_event_info, ev, old_off, logname);
+    Exit_status retval= process_event(print_event_info, ev, old_off,
+                                      old_off + (*len - 1), logname);
     if (retval != OK_CONTINUE)
       DBUG_RETURN(retval);
   }
@@ -2397,7 +2402,8 @@ static Exit_status handle_event_text_mode(PRINT_EVENT_INFO *print_event_info,
       DBUG_RETURN(ERROR_STOP);
     }
 
-    retval= process_event(print_event_info, ev, old_off, logname);
+    retval= process_event(print_event_info, ev, old_off, old_off + (*len - 1),
+                          logname);
     if (retval != OK_CONTINUE)
     {
       my_close(file,MYF(MY_WME));
@@ -2796,6 +2802,7 @@ static Exit_status check_header(IO_CACHE* file,
           */
           Exit_status retval= process_event(print_event_info,
                                             new_description_event, tmp_pos,
+                                            my_b_tell(file),
                                             logname);
           if (retval != OK_CONTINUE)
             return retval;
@@ -2944,22 +2951,12 @@ static Exit_status dump_local_log_entries(PRINT_EVENT_INFO *print_event_info,
       }
       // else file->error == 0 means EOF, that's OK, we break in this case
 
-      /*
-        Emit a warning in the event that we finished processing input
-        before reaching the boundary indicated by --stop-position.
-      */
-      if (((longlong)stop_position != stop_position_default) &&
-          stop_position > my_b_tell(file))
-      {
-          retval = OK_STOP;
-          warning("Did not reach stop position %llu before "
-                  "end of input", stop_position);
-      }
 
       goto end;
     }
-    if ((retval= process_event(print_event_info, ev, old_off, logname)) !=
-        OK_CONTINUE)
+    if ((retval= process_event(print_event_info, ev, old_off, my_b_tell(file),
+                               logname)) != OK_CONTINUE)
+
       goto end;
   }
 
@@ -3135,9 +3132,17 @@ int main(int argc, char** argv)
 
   if (stop_datetime != MY_TIME_T_MAX &&
       stop_datetime > last_processed_datetime)
-      warning("Did not reach stop datetime '%s' "
-              "before end of input", stop_datetime_str);
+    warning("Did not reach stop datetime '%s' before end of input",
+            stop_datetime_str);
 
+  /*
+    Emit a warning in the event that we finished processing input
+    before reaching the boundary indicated by --stop-position.
+  */
+  if (((longlong) stop_position != stop_position_default) &&
+      stop_position > last_processed_position)
+    warning("Did not reach stop position %llu before end of input",
+            stop_position);
   /*
     If enable flashback, need to print the events from the end to the
     beginning
diff --git a/mysql-test/suite/binlog/r/binlog_mysqlbinlog_warn_stop_position.result b/mysql-test/suite/binlog/r/binlog_mysqlbinlog_warn_stop_position.result
index 1517fd765ac..40fb791cf02 100644
--- a/mysql-test/suite/binlog/r/binlog_mysqlbinlog_warn_stop_position.result
+++ b/mysql-test/suite/binlog/r/binlog_mysqlbinlog_warn_stop_position.result
@@ -18,6 +18,51 @@ drop table t1;
 # Ensuring file offset of binlog_f2_mid < binlog_f1_end
 #
 #
+# Test using --read-from-remote-server
+#
+connection default;
+#
+# --stop-position tests
+#
+#  Case 1.a) With one binlog file, a --stop-position before the end of
+# the file should not result in a warning
+# MYSQL_BINLOG --read-from-remote-server --stop-position=binlog_f1_pre_rotate binlog_f1_full --result-file=tmp/warn_position_test_file.out 2>&1
+#
+#  Case 1.b) With one binlog file, a --stop-position at the exact end of
+# the file should not result in a warning
+# MYSQL_BINLOG --read-from-remote-server --stop-position=binlog_f1_end binlog_f1_full --result-file=tmp/warn_position_test_file.out 2>&1
+#
+#  Case 1.c) With one binlog file, a --stop-position past the end of the
+# file should(!) result in a warning
+# MYSQL_BINLOG --read-from-remote-server --short-form --stop-position=binlog_f1_over_eof binlog_f1_full --result-file=tmp/warn_position_test_file.out 2>&1
+WARNING: Did not reach stop position <BINLOG_F1_OVER_EOF> before end of input
+#
+#  Case 2.a) With two binlog files, a --stop-position targeting b2 which
+# exists in the size of b1 should:
+#    1) not provide any warnings
+#    2) not prevent b2 from outputting its desired events before the
+#       stop position
+# MYSQL_BINLOG --read-from-remote-server --stop-position=binlog_f2_mid binlog_f1_full binlog_f2_full --result-file=tmp/warn_position_test_file.out 2>&1
+include/assert_grep.inc [Ensure all intended GTIDs are present]
+include/assert_grep.inc [Ensure the next GTID binlogged is _not_ present]
+#
+#  Case 2.b) With two binlog files, a --stop-position targeting the end
+# of binlog 2 should:
+#    1) not provide any warnings
+#    2) not prevent b2 from outputting its entire binary log
+# MYSQL_BINLOG --read-from-remote-server --stop-position=binlog_f2_end binlog_f1_full binlog_f2_full --result-file=tmp/warn_position_test_file.out 2>&1
+include/assert_grep.inc [Ensure a GTID exists for each transaction]
+include/assert_grep.inc [Ensure the last GTID binlogged is present]
+#
+#  Case 2.c) With two binlog files, a --stop-position targeting beyond
+# the eof of binlog 2 should:
+#    1) provide a warning that the stop position was not reached
+#    2) not prevent b2 from outputting its entire binary log
+# MYSQL_BINLOG --read-from-remote-server --stop-position=binlog_f2_over_eof binlog_f1_full binlog_f2_full --result-file=tmp/warn_position_test_file.out 2>&1
+WARNING: Did not reach stop position <BINLOG_F2_OVER_EOF> before end of input
+include/assert_grep.inc [Ensure a GTID exists for each transaction]
+#
+#
 # Test using local binlog files
 #
 connection default;
diff --git a/mysql-test/suite/binlog/t/binlog_mysqlbinlog_warn_stop_position.inc b/mysql-test/suite/binlog/t/binlog_mysqlbinlog_warn_stop_position.inc
index 7ecf5c1098a..878de8ef2b5 100644
--- a/mysql-test/suite/binlog/t/binlog_mysqlbinlog_warn_stop_position.inc
+++ b/mysql-test/suite/binlog/t/binlog_mysqlbinlog_warn_stop_position.inc
@@ -50,7 +50,7 @@ if ($use_local_files)
 --echo #
 --echo #  Case 1.c) With one binlog file, a --stop-position past the end of the
 --echo # file should(!) result in a warning
---let $binlog_f1_over_eof= `SELECT $binlog_f1_end + 10`
+--let $binlog_f1_over_eof= `SELECT $binlog_f1_end + 1`
 --echo # MYSQL_BINLOG $PARAM_READ_FROM_REMOTE_OUT --short-form --stop-position=binlog_f1_over_eof binlog_f1_full --result-file=$binlog_out_relpath 2>&1
 --replace_result $binlog_f1_over_eof <BINLOG_F1_OVER_EOF>
 --exec $MYSQL_BINLOG $PARAM_READ_FROM_REMOTE --short-form --stop-position=$binlog_f1_over_eof $binlog_f1_full --result-file=$binlog_out 2>&1
@@ -102,7 +102,7 @@ if ($use_local_files)
 --echo # the eof of binlog 2 should:
 --echo #    1) provide a warning that the stop position was not reached
 --echo #    2) not prevent b2 from outputting its entire binary log
---let $binlog_f2_over_eof= `SELECT $binlog_f2_end + 10`
+--let $binlog_f2_over_eof= `SELECT $binlog_f2_end + 1`
 --echo # MYSQL_BINLOG $PARAM_READ_FROM_REMOTE_OUT --stop-position=binlog_f2_over_eof binlog_f1_full binlog_f2_full --result-file=$binlog_out_relpath 2>&1
 --replace_result $binlog_f2_over_eof <BINLOG_F2_OVER_EOF>
 --exec $MYSQL_BINLOG $PARAM_READ_FROM_REMOTE --stop-position=$binlog_f2_over_eof $binlog_f1_full $binlog_f2_full --result-file=$binlog_out 2>&1
diff --git a/mysql-test/suite/binlog/t/binlog_mysqlbinlog_warn_stop_position.test b/mysql-test/suite/binlog/t/binlog_mysqlbinlog_warn_stop_position.test
index c1028a23404..34ec9e67579 100644
--- a/mysql-test/suite/binlog/t/binlog_mysqlbinlog_warn_stop_position.test
+++ b/mysql-test/suite/binlog/t/binlog_mysqlbinlog_warn_stop_position.test
@@ -64,13 +64,12 @@ if ($binlog_f2_mid > $binlog_f1_end)
   --die Mid point chosen to end in binlog 2 does not exist in earlier binlog
 }
 
-#--echo #
-#--echo #
-#--echo # Test using --read-from-remote-server
-#--echo #
-#--let $use_local_files= 0
-#--emit warning is not supported by --read-from-remote-server now
-#--source binlog_mysqlbinlog_warn_stop_position.inc
+--echo #
+--echo #
+--echo # Test using --read-from-remote-server
+--echo #
+--let $use_local_files= 0
+--source binlog_mysqlbinlog_warn_stop_position.inc
 
 --echo #
 --echo #
```